js-core-3/week-2/lesson.md

```diff
@@ -1,305 +1,302 @@
-   understand what the function's purpose is and how each part of the code
-   contributes to it. By writing comments, the coder communicates their
-   reasoning and helps the function be human readable.
+### Exercise
 
-3. Layout/formatting: unnecessary spacing between the if and else statement.
+#### Check-in
 
-4. Single responsibility: the function doesn't have a single purpose. It
-   calculates national insurance and salary deductions. Maybe the national
-   insurance calculation could be moved to a separate function.
+Here is an example of Spotify's REST endpoint to allow you to get the tracks from a playlist
 
-5. Input variable being overwritten: the function requires gross salary (before
-   deductions) and net salary (after deductions) the `salary` input variable is
-   therefore copied into an `originalSalary` variable so that it can be changed.
-   It would be much clearer to create a new `netSalary` variable and leave
-   `salary` unmodified.
+    GET https://api.spotify.com/v1/playlists/{playlist_id}/tracks
 
-6. DRY principle: the function validates the DRY (Don't Repeat Yourself) rule.
-   The line where a deduction is taken from the salary is repeated 3 times with
-   different indices. This can be replaced with a `for` loop.
+**Task (5 mins):**
 
-7. Magic numbers. The code contains a lot of magic numbers, including `17775`,
-   `0.09` and `0.1`.
+    On Instagram, you can access a post via the following endpoint:
 
-8. Useless parameters: the code contains a variable which isn't used. They
-   should be removed because they are confusing. It is tempting when you're
-   starting to code a function to add more parameters thinking that you might
-   need them, but it's important to remove them if you don't end up using them.
+    https://api.instagram.com/v1/media/{media-id}
 
-> Exercise: Working in pairs, go through all of these issues and make
-> appropriate improvements to the code.
+    1. Write the URL of the endpoint to get the list of comments on a given instagram post?
+    2. What HTTP verb will you use to make requests to this endpoint?
 
-# Async vs Sync
-ToDO: Trace async code on paper
+## 2. How to use the `fetch` API to do AJAX calls
 
-- Different ways of doing async in JavaScript
+### Explanation
 
+- Modern replacement of XMLHttpRequest
+- Uses Promise structure
+- The Fetch API is defined in the browser (window.fetch)
+- Only modern browsers support it (show [caniuse.com](https://caniuse.com/#feat=fetch))
+- Fetch API documentations by mozilla [link](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch)
 
-# Intro to ES6
+### What is Fetch made of?
 
-ECMAScript 2015 (or ES6) is a significant update to JavaScript that introduces
-new syntax for writing complex applications. 
+Fetch is an example of a javascript which has a very specific structure.
 
-## const and let
+```javascript
+let promiseToDoHomeWork = new Promise(function (resolve, reject) {
```

Comment:
How about a slightly less artificial example?
```javascript
let clickCount;

document.querySelector("button").addEventListener("click", () => clickCount++);

let promiseOfClicks = new Promise(function(resolve, reject) {
  clickCount = 0;
  console.log("Counting clicks for 5 seconds...");
  setTimeout(() => {
    if (clickCount > 0) {
      resolve(clickCount);
    } else {
      reject();
    }
  }, 5000)
});

promiseOfClicks
  .then(clicks => console.log(`Clicked ${clicks} times`)
  .catch(error => console.warn(error));
```

It's still pretty artificial, but at least it passes around _some_ data...

I worry that when we use entirely contrived examples, the students often don't get the point...